File: output/java/tests/ternary_expr.java
import java.util.*;


import java.util.*;

// Test: Ternary/conditional expressions in Frame handlers
// Java uses: cond ? a : b

class TernaryTestFrameEvent {
    String _message;
    HashMap<String, Object> _parameters;

    TernaryTestFrameEvent(String message) {
        this._message = message;
        this._parameters = null;
    }

    TernaryTestFrameEvent(String message, HashMap<String, Object> parameters) {
        this._message = message;
        this._parameters = parameters;
    }
}

class TernaryTestFrameContext {
    TernaryTestFrameEvent _event;
    Object _return;
    HashMap<String, Object> _data;

    TernaryTestFrameContext(TernaryTestFrameEvent event, Object defaultReturn) {
        this._event = event;
        this._return = defaultReturn;
        this._data = new HashMap<>();
    }
}

class TernaryTestCompartment {
    String state;
    HashMap<String, Object> state_args;
    HashMap<String, Object> state_vars;
    HashMap<String, Object> enter_args;
    HashMap<String, Object> exit_args;
    TernaryTestFrameEvent forward_event;
    TernaryTestCompartment parent_compartment;

    TernaryTestCompartment(String state) {
        this.state = state;
        this.state_args = new HashMap<>();
        this.state_vars = new HashMap<>();
        this.enter_args = new HashMap<>();
        this.exit_args = new HashMap<>();
        this.forward_event = null;
        this.parent_compartment = null;
    }

    TernaryTestCompartment copy() {
        TernaryTestCompartment c = new TernaryTestCompartment(this.state);
        c.state_args = new HashMap<>(this.state_args);
        c.state_vars = new HashMap<>(this.state_vars);
        c.enter_args = new HashMap<>(this.enter_args);
        c.exit_args = new HashMap<>(this.exit_args);
        c.forward_event = this.forward_event;
        c.parent_compartment = this.parent_compartment;
        return c;
    }
}

class TernaryTest {
    private ArrayList<TernaryTestCompartment> _state_stack;
    private TernaryTestCompartment __compartment;
    private TernaryTestCompartment __next_compartment;
    private ArrayList<TernaryTestFrameContext> _context_stack;
    public boolean cond = true;

    public TernaryTest() {
        _state_stack = new ArrayList<>();
        _context_stack = new ArrayList<>();
        __compartment = new TernaryTestCompartment("Ready");
        __next_compartment = null;
        TernaryTestFrameEvent __frame_event = new TernaryTestFrameEvent("$>");
        TernaryTestFrameContext __ctx = new TernaryTestFrameContext(__frame_event, null);
        _context_stack.add(__ctx);
        __kernel(_context_stack.get(_context_stack.size() - 1)._event);
        _context_stack.remove(_context_stack.size() - 1);
    }

    private void __kernel(TernaryTestFrameEvent __e) {
        __router(__e);
        while (__next_compartment != null) {
            var next_compartment = __next_compartment;
            __next_compartment = null;
            TernaryTestFrameEvent exit_event = new TernaryTestFrameEvent("<$");
            __router(exit_event);
            __compartment = next_compartment;
            if (__compartment.forward_event == null) {
                TernaryTestFrameEvent enter_event = new TernaryTestFrameEvent("$>");
                __router(enter_event);
            } else {
                var forward_event = __compartment.forward_event;
                __compartment.forward_event = null;
                if (forward_event._message.equals("$>")) {
                    __router(forward_event);
                } else {
                    TernaryTestFrameEvent enter_event = new TernaryTestFrameEvent("$>");
                    __router(enter_event);
                    __router(forward_event);
                }
            }
        }
    }

    private void __router(TernaryTestFrameEvent __e) {
        String state_name = __compartment.state;
        if (state_name.equals("Ready")) {
            _state_Ready(__e);
        }
    }

    private void __transition(TernaryTestCompartment next) {
        __next_compartment = next;
    }

    public int get_value() {
        TernaryTestFrameEvent __e = new TernaryTestFrameEvent("get_value");
        TernaryTestFrameContext __ctx = new TernaryTestFrameContext(__e, null);
        _context_stack.add(__ctx);
        __kernel(_context_stack.get(_context_stack.size() - 1)._event);
        var __result = (int) _context_stack.get(_context_stack.size() - 1)._return;
        _context_stack.remove(_context_stack.size() - 1);
        return __result;
    }

    public void set_condition(boolean c) {
        HashMap<String, Object> __params = new HashMap<>();
        __params.put("c", c);
        TernaryTestFrameEvent __e = new TernaryTestFrameEvent("set_condition", __params);
        TernaryTestFrameContext __ctx = new TernaryTestFrameContext(__e, null);
        _context_stack.add(__ctx);
        __kernel(_context_stack.get(_context_stack.size() - 1)._event);
        _context_stack.remove(_context_stack.size() - 1);
    }

    private void _state_Ready(TernaryTestFrameEvent __e) {
        if (__e._message.equals("get_value")) {
            int result = this.cond ? 100 : 200;
            _context_stack.get(_context_stack.size() - 1)._return = result;
        } else if (__e._message.equals("set_condition")) {
            var c = (boolean) __e._parameters.get("c");
            this.cond = c;
        }
    }
}

class Main {
    public static void main(String[] args) {
        System.out.println("TAP version 14");
        System.out.println("1..2");

        TernaryTest s = new TernaryTest();

        // cond=true: should return 100
        int v1 = (int) s.get_value();
        if (v1 == 100) {
            System.out.println("ok 1 - cond=true returns 100");
        } else {
            System.out.println("not ok 1 - cond=true returns 100 # got " + v1);
        }

        // cond=false: should return 200
        s.set_condition(false);
        int v2 = (int) s.get_value();
        if (v2 == 200) {
            System.out.println("ok 2 - cond=false returns 200");
        } else {
            System.out.println("not ok 2 - cond=false returns 200 # got " + v2);
        }
    }
}

File: output/cpp/tests/ternary_expr.cpp
#include <string>
#include <unordered_map>
#include <vector>
#include <any>
#include <memory>
#include <functional>


// Test: Ternary/conditional expressions in Frame handlers
// C++ uses: cond ? a : b

#include <iostream>
#include <string>
#include <cassert>

class TernaryTestFrameEvent {
public:
    std::string _message;
    std::unordered_map<std::string, std::any> _parameters;

    TernaryTestFrameEvent(const std::string& message, std::unordered_map<std::string, std::any> params = {})
        : _message(message), _parameters(std::move(params)) {}
};

class TernaryTestFrameContext {
public:
    TernaryTestFrameEvent _event;
    std::any _return;
    std::unordered_map<std::string, std::any> _data;

    TernaryTestFrameContext(TernaryTestFrameEvent event, std::any default_return = {})
        : _event(std::move(event)), _return(std::move(default_return)) {}
};

class TernaryTestCompartment {
public:
    std::string state;
    std::unordered_map<std::string, std::any> state_args;
    std::unordered_map<std::string, std::any> state_vars;
    std::unordered_map<std::string, std::any> enter_args;
    std::unordered_map<std::string, std::any> exit_args;
    std::unique_ptr<TernaryTestFrameEvent> forward_event;
    std::unique_ptr<TernaryTestCompartment> parent_compartment;

    explicit TernaryTestCompartment(const std::string& state) : state(state) {}

    std::unique_ptr<TernaryTestCompartment> clone() const {
        auto c = std::make_unique<TernaryTestCompartment>(state);
        c->state_args = state_args;
        c->state_vars = state_vars;
        c->enter_args = enter_args;
        c->exit_args = exit_args;
        return c;
    }
};

class TernaryTest {
private:
    std::vector<std::unique_ptr<TernaryTestCompartment>> _state_stack;
    std::unique_ptr<TernaryTestCompartment> __compartment;
    std::unique_ptr<TernaryTestCompartment> __next_compartment;
    std::vector<TernaryTestFrameContext> _context_stack;

    void __kernel(TernaryTestFrameEvent& __e) {
        __router(__e);
        while (__next_compartment) {
            auto next_compartment = std::move(__next_compartment);
            TernaryTestFrameEvent exit_event("<$");
            __router(exit_event);
            __compartment = std::move(next_compartment);
            if (!__compartment->forward_event) {
                TernaryTestFrameEvent enter_event("$>");
                __router(enter_event);
            } else {
                auto forward_event = std::move(__compartment->forward_event);
                if (forward_event->_message == "$>") {
                    __router(*forward_event);
                } else {
                    TernaryTestFrameEvent enter_event("$>");
                    __router(enter_event);
                    __router(*forward_event);
                }
            }
        }
    }

    void __router(TernaryTestFrameEvent& __e) {
        const auto& state_name = __compartment->state;
        if (state_name == "Ready") {
            _state_Ready(__e);
        }
    }

    void __transition(std::unique_ptr<TernaryTestCompartment> next) {
        __next_compartment = std::move(next);
    }

    void _state_Ready(TernaryTestFrameEvent& __e) {
        if (__e._message == "get_value") {
            auto result = cond ? 100 : 200;
            _context_stack.back()._return = std::any(result);
        } else if (__e._message == "set_condition") {
            auto c = std::any_cast<bool>(__e._parameters.at("c"));
            cond = c;
        }
    }

public:
    bool cond = true;

    TernaryTest() {
        __compartment = std::make_unique<TernaryTestCompartment>("Ready");
        TernaryTestFrameEvent __frame_event("$>");
        TernaryTestFrameContext __ctx(std::move(__frame_event));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        _context_stack.pop_back();
    }

    int get_value() {
        TernaryTestFrameEvent __e("get_value");
        TernaryTestFrameContext __ctx(std::move(__e), std::any(int()));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        auto __result = std::any_cast<int>(std::move(_context_stack.back()._return));
        _context_stack.pop_back();
        return __result;
    }

    void set_condition(bool c) {
        std::unordered_map<std::string, std::any> __params;
        __params["c"] = c;
        TernaryTestFrameEvent __e("set_condition", std::move(__params));
        TernaryTestFrameContext __ctx(std::move(__e));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        _context_stack.pop_back();
    }
};

int main() {
    printf("TAP version 14\n");
    printf("1..2\n");

    TernaryTest s;

    // cond=true: should return 100
    auto v1 = s.get_value();
    if (v1 == 100) {
        printf("ok 1 - cond=true returns 100\n");
    } else {
        printf("not ok 1 - cond=true returns 100 # got %d\n", v1);
    }

    // cond=false: should return 200
    s.set_condition(false);
    auto v2 = s.get_value();
    if (v2 == 200) {
        printf("ok 2 - cond=false returns 200\n");
    } else {
        printf("not ok 2 - cond=false returns 200 # got %d\n", v2);
    }

    return 0;
}
